test: enable running independent functional test sub-tests

ismaelsadeeq opened the pull request:
- Some test methods in the functional test framework are independent and do not require any prior context or setup in `run_test`.
- This commit adds a new option for running these specific methods within a test file, allowing them to be executed individually without running the entire test suite.
- Using this option reduces the time you need to wait before the test you are interested in starts executing.
- The functionality added by this PR can be achieved manually by commenting out code, but having a pragmatic option to do this is more convenient.

Note: Running test methods that require arguments or context will fail.

**Example Usage**:
```zsh
build/test/functional/feature_reindex.py --test_methods continue_reindex_after_shutdown
```

```zsh
build/test/functional/feature_config_args.py --test_methods test_log_buffer test_args_log test_connect_with_seednode
```

rkrux commented on test/functional/test_framework/test_framework.py:
```diff
@@ -155,6 +159,15 @@ def main(self):
             exit_code = self.shutdown()
             sys.exit(exit_code)
 
+    def run_custom_test(self):
+        for method_name in self.options.test_methods:
```
Since the argument is a list named `test_methods`, let's call this function `run_custom_tests`?

ismaelsadeeq replied:
Done, thanks for reviewing

rkrux commented on test/functional/test_framework/test_framework.py:
```diff
@@ -155,6 +159,15 @@ def main(self):
+    def run_custom_test(self):
+        for method_name in self.options.test_methods:
+            method = getattr(self, method_name, None)
+            if not method:
+                raise AttributeError(f"No method with name {method_name}")
+            if not callable(method):
+                raise TypeError(f"{method_name} is not callable")
```
Might as well find the searchable & callable methods initially and execute them while throwing the error for those that are filtered out? Otherwise few would be executed and the first incorrect one would throw error.

ismaelsadeeq replied:
Done, thanks

rkrux commented on test/functional/test_framework/test_framework.py:
```diff
@@ -194,6 +207,8 @@ def parse_args(self, test_file):
                             help="use BIP324 v2 connections between all nodes by default")
         parser.add_argument("--v1transport", dest="v1transport", default=False, action="store_true",
                             help="Explicitly use v1 transport (can be used to overwrite global --v2transport option)")
+        parser.add_argument("--test_methods", dest="test_methods", nargs='*',
+                            help="Run specified test methods sequentially instead of the full test. Use only for methods that do not depend on any context set up in run_test or other methods.")
```
The PR title states that this will enable running functional test sub-tests but any method can be run with this argument at the moment such as one below. Should we restrict only for those that have the `test_` prefix?

```
+    def sample_calculator(self):
+        self.log.info("Sample calculator running")
+        x = 2 + 3
+        self.log.info(f'Calculated x to be: {x}')
+
```

ismaelsadeeq replied:
Some tests might not have the test prefix, though. I don’t think we can use the `test_` prefix to catch all sub-tests; rather, the goal is just to enable running individual methods that can be sub-tests.

rkrux replied:
> or other methods.

Nit: This can be reworded though because the following snippet works.

```
    def calculate(self, val):
        return val

    def sample_calculator(self):
        self.log.info("Sample calculator running")
        x = self.calculate(2 + 3)
        self.log.info(f'Calculated x to be: {x}')
```

```
➜  bitcoin git:(09-2024-run-ind-functional-test) ✗ build/test/functional/feature_config_args.py --test_methods test_log_buffer test_args_log sample_calculator
```

ismaelsadeeq replied:
leaving this as is; but happy to take explicit suggestions :)

rkrux replied:
If I'm not missing something, why not just remove it?

ismaelsadeeq replied:
It is useful to because running `build/test/functional/test_runner.py --help`, will describe this new feature for users. If this is removed, it will not be documented, and future users may not learn about it.
Therefore, I will be leaving it as is.

rkrux replied:
I meant removing only "or other methods".

ismaelsadeeq replied:
Happy to update when retouching.

rkrux reviewed (approved):
Concept ACK cc21876b125930f8320dbb95016f9ee7c1ffec55

Successful make and functional tests. Tried running custom methods with the new `--test_methods` argument.
Left couple suggestions.

ismaelsadeeq committed:
test: enable running individual independent functional test methods

- Some test methods in the functional test framework are independent
  and do not require any previous context or setup defined in `run_test`.
- This commit adds a new option for running these specific methods within a test file,
  allowing them to be executed individually without running the entire test suite.

- running test methods that require an argument or context will fail.

rkrux reviewed (approved):
ACK e96823bdcc2317fbf0bed64e53a1b3e5bb8f00f6

Thanks for quickly addressing the feedback.

ismaelsadeeq commented on test/functional/test_framework/test_framework.py:
```diff
@@ -155,6 +159,28 @@ def main(self):
+        for method in correct_methods:
+            method()
+
+        if incorrect_methods or uncallable_methods:
+            error_message = f"Errors encountered:"
+            if incorrect_methods:
+                error_message += f"\nIncorrect methods (not found): {incorrect_methods}"
+            if uncallable_methods:
+                error_message += f"\nUncallable methods: {uncallable_methods}"
+            raise RuntimeError(error_message)
```
The approach here was added to address [this comment](https://github.com/bitcoin/bitcoin/pull/30991#discussion_r1827274569) cc @rkrux .

Which is an overkill for this simple feature honestly.

Is the simplified approach below preferable?

```python3
def run_custom_tests(self):
    for method_name in self.options.test_methods:
        self.log.info(f"Attempting to execute method: {method_name}")
        method = getattr(self, method_name)
        method()
        self.log.info(f"Method '{method_name}' executed successfully.")
```

This version has a few more advantage. 
- It logs each method's execution attempt and relies on Python’s built-in exceptions to handle issues.
-  The code is simpler. 
- Additionally, it provides helpful feedback if a method name is close to a valid method but has a typo. For instance, omitting the final "n" in the method name would result in:
when running the example in the OP will result in

```
Traceback (most recent call last):
  File "/Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/test/functional/test_framework/test_framework.py", line 133, in main
    self.run_custom_tests()
  File "/Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/test/functional/test_framework/test_framework.py", line 165, in run_custom_tests
    method = getattr(self, method_name)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'ReindexTest' object has no attribute 'continue_reindex_after_shutdow'. Did you mean: 'continue_reindex_after_shutdown'?
```

rkrux replied:
Since [this commit](https://github.com/bitcoin/bitcoin/commit/cc21876b125930f8320dbb95016f9ee7c1ffec55) was checking for the presence and call-ability of the passed methods already, the main intent of [this comment](https://github.com/bitcoin/bitcoin/pull/30991#discussion_r1827274569) was to add on to it by allowing the execution of all the methods in case of failure of any one of them.

> test-only dev-only debug-only feature

I believe keeping this^ in mind, it makes sense to just throw the error in case if any invalid method is passed. I agree the approach shared in the above comment is simpler and good enough for the use case.

Though, I feel sticking to only one term between `test_methods` or `custom_tests` would be better from readability POV.

rkrux replied:
With the above approach, I see the following outputs that seem okay.

Error when a non existing attribute is passed.
```
2024-11-11T06:12:11.578000Z TestFramework (INFO): Attempting to execute method: test_args_l
2024-11-11T06:12:11.579000Z TestFramework (ERROR): Unexpected exception caught during testing
Traceback (most recent call last):
  File "/Users/rkrux/projects/bitcoin/test/functional/test_framework/test_framework.py", line 133, in main
    self.run_custom_tests()
    ~~~~~~~~~~~~~~~~~~~~~^^
  File "/Users/rkrux/projects/bitcoin/test/functional/test_framework/test_framework.py", line 165, in run_custom_tests
    method = getattr(self, method_name)
AttributeError: 'ConfArgsTest' object has no attribute 'test_args_l'. Did you mean: 'test_args_log'?
```

Error when a non callable attribute is passed.
```
2024-11-11T06:16:15.078000Z TestFramework (INFO): Attempting to execute method: chain
2024-11-11T06:16:15.078000Z TestFramework (ERROR): Unexpected exception caught during testing
Traceback (most recent call last):
  File "/Users/rkrux/projects/bitcoin/test/functional/test_framework/test_framework.py", line 133, in main
    self.run_custom_tests()
    ~~~~~~~~~~~~~~~~~~~~~^^
  File "/Users/rkrux/projects/bitcoin/test/functional/test_framework/test_framework.py", line 166, in run_custom_tests
    method()
    ~~~~~~^^
TypeError: 'str' object is not callable
```

ismaelsadeeq replied:
Done in latest push.
thanks

rkrux reviewed (approved):
reACK 409d0d629378c3e23388ed31516376ad1ae536b5

ryanofsky reviewed (approved):
Code review ACK 409d0d629378c3e23388ed31516376ad1ae536b5. This seems like a good step towards making it easy to run independent tests quickly. I think ideally there would be some naming convention or @ annotation added to test methods that can run independently, so the test framework could provide more functionality like being able to list test methods, being able to show command lines to quickly reproduce problems when tests fails, and calling test methods automatically instead of requiring individual tests to call them. But these ideas are all compatible with the new `--test_methods` option
